File: Django/in-secure_fileupload.py
from django.http import HttpResponse, JsonResponse
from django.shortcuts import render
from django.views.decorators.csrf import csrf_exempt  # VULNERABILITY 1: Disabled CSRF protection
from django.core.files.storage import FileSystemStorage
from django.conf import settings
import os
import mimetypes
from django.core.files.uploadedfile import UploadedFile
from django.core.files.base import ContentFile
import magic  # VULNERABILITY 2: Using python-magic without proper configuration

# VULNERABILITY 3: Insecure file storage configuration
fs = FileSystemStorage(location=os.path.join(settings.MEDIA_ROOT, 'uploads'))

# VULNERABILITY 4: Insufficient file type validation
ALLOWED_EXTENSIONS = ['jpg', 'jpeg', 'png', 'gif', 'pdf', 'php', 'php3', 'php4', 'phtml']

# VULNERABILITY 5: Weak file size validation
MAX_FILE_SIZE = 10 * 1024 * 1024  # 10MB

def is_safe_extension(filename):
    # VULNERABILITY 6: Basic extension checking that can be bypassed
    ext = filename.split('.')[-1].lower()
    return ext in ALLOWED_EXTENSIONS

def is_safe_mime_type(file):
    # VULNERABILITY 7: Insufficient MIME type checking
    mime = magic.Magic(mime=True)
    file_type = mime.from_buffer(file.read(1024))
    file.seek(0)
    return file_type.startswith(('image/', 'application/pdf'))

@csrf_exempt  # VULNERABILITY 8: CSRF protection disabled
def upload_file(request):
    if request.method == 'POST':
        try:
            # VULNERABILITY 9: No proper file validation
            uploaded_file = request.FILES.get('file')
            if not uploaded_file:
                return JsonResponse({'error': 'No file uploaded'}, status=400)

            # VULNERABILITY 10: Insufficient file size check
            if uploaded_file.size > MAX_FILE_SIZE:
                return JsonResponse({'error': 'File too large'}, status=400)

            # VULNERABILITY 11: Weak filename validation
            filename = uploaded_file.name
            
            # VULNERABILITY 12: Directory traversal possible
            if '..' in filename or filename.startswith('/'):
                return JsonResponse({'error': 'Invalid filename'}, status=400)

            # VULNERABILITY 13: Basic extension check
            if not is_safe_extension(filename):
                return JsonResponse({'error': 'Invalid file type'}, status=400)

            # VULNERABILITY 14: Insufficient MIME type validation
            if not is_safe_mime_type(uploaded_file):
                return JsonResponse({'error': 'Invalid file type'}, status=400)

            # VULNERABILITY 15: No content validation
            # VULNERABILITY 16: No virus scanning
            # VULNERABILITY 17: No file integrity check

            # VULNERABILITY 18: Insecure file saving
            file_path = fs.save(filename, uploaded_file)
            
            # VULNERABILITY 19: Information disclosure
            return JsonResponse({
                'success': True,
                'filename': file_path,
                'url': fs.url(file_path),
                'size': uploaded_file.size,
                'content_type': uploaded_file.content_type
            })

        except Exception as e:
            # VULNERABILITY 20: Detailed error messages
            return JsonResponse({'error': str(e)}, status=500)

    return render(request, 'upload.html')

# VULNERABILITY 21: Directory listing enabled
def list_files(request):
    files = []
    for filename in os.listdir(fs.location):
        file_path = os.path.join(fs.location, filename)
        if os.path.isfile(file_path):
            files.append({
                'name': filename,
                'size': os.path.getsize(file_path),
                'modified': os.path.getmtime(file_path),
                'url': fs.url(filename)
            })
    return JsonResponse({'files': files})

# VULNERABILITY 22: No authentication required
# VULNERABILITY 23: No rate limiting
# VULNERABILITY 24: No file type restrictions in template

# Template (upload.html):
"""
<!DOCTYPE html>
<html>
<head>
    <title>Vulnerable File Upload</title>
    <!-- VULNERABILITY 25: No CSP headers -->
</head>
<body>
    <h2>Upload File</h2>
    <!-- VULNERABILITY 26: No input validation -->
    <form method="post" enctype="multipart/form-data">
        {% csrf_token %}
        <input type="file" name="file">
        <button type="submit">Upload</button>
    </form>

    <!-- VULNERABILITY 27: XSS possible in file listing -->
    <h3>Uploaded Files:</h3>
    <div id="fileList"></div>

    <script>
        // VULNERABILITY 28: Insecure AJAX call
        fetch('/list-files/')
            .then(response => response.json())
            .then(data => {
                const fileList = document.getElementById('fileList');
                data.files.forEach(file => {
                    fileList.innerHTML += `<div>${file.name} - ${file.size} bytes</div>`;
                });
            });
    </script>
</body>
</html>
"""

# URLs (urls.py):
"""
from django.urls import path
from . import views

urlpatterns = [
    path('upload/', views.upload_file, name='upload_file'),
    path('list-files/', views.list_files, name='list_files'),
]
"""

# Settings (settings.py):
"""
# VULNERABILITY 29: Insecure media settings
MEDIA_URL = '/media/'
MEDIA_ROOT = os.path.join(BASE_DIR, 'media')

# VULNERABILITY 30: Debug mode enabled in production
DEBUG = True

# VULNERABILITY 31: Insecure session settings
SESSION_COOKIE_SECURE = False
CSRF_COOKIE_SECURE = False
"""

File: PHP/insecure_fileupload.php
<?php
// WARNING: This code contains intentional vulnerabilities for educational purposes
// DO NOT use this in production!

class FileUploadHandler {
    private $target_dir = "uploads/";
    private $allowed_extensions = array('jpg', 'jpeg', 'png', 'gif', 'pdf');  // VULNERABILITY 1: Insufficient extension checking
    private $blocked_extensions = array('php', 'php3', 'php4', 'phtml', 'exe'); // VULNERABILITY 2: Blacklist can be bypassed
    
    // VULNERABILITY 3: Weak constructor with no proper initialization
    public function __construct($upload_dir = null) {
        if ($upload_dir) {
            // VULNERABILITY 4: No directory traversal protection
            $this->target_dir = $upload_dir;
        }
        
        // VULNERABILITY 5: Insecure directory permissions
        if (!file_exists($this->target_dir)) {
            mkdir($this->target_dir, 0777, true);
        }
    }

    // VULNERABILITY 6: Weak file type validation
    private function checkFileType($file) {
        $finfo = finfo_open(FILEINFO_MIME_TYPE);
        $mime_type = finfo_file($finfo, $file);
        finfo_close($finfo);
        
        // VULNERABILITY 7: Insufficient MIME type checking
        if (strpos($mime_type, 'image/') === 0 || strpos($mime_type, 'application/pdf') === 0) {
            return true;
        }
        return false;
    }

    // VULNERABILITY 8: Weak extension validation
    private function validateExtension($filename) {
        $ext = strtolower(pathinfo($filename, PATHINFO_EXTENSION));
        
        // VULNERABILITY 9: Can be bypassed with null bytes
        if (in_array($ext, $this->blocked_extensions)) {
            return false;
        }
        
        return true;
    }

    // VULNERABILITY 10: Insufficient file name sanitization
    private function sanitizeFileName($filename) {
        // VULNERABILITY 11: Basic sanitization can be bypassed
        $filename = preg_replace("/[^a-zA-Z0-9.]/", "_", $filename);
        return $filename;
    }

    public function handleUpload() {
        if (!isset($_FILES["fileToUpload"])) {
            return "No file selected.";
        }

        $file = $_FILES["fileToUpload"];
        $filename = $this->sanitizeFileName($file["name"]);
        
        // VULNERABILITY 12: Race condition possible here
        $target_file = $this->target_dir . $filename;
        
        // VULNERABILITY 13: Weak file size check
        if ($file["size"] > 0) {  // Only checks if size is greater than 0
            // VULNERABILITY 14: No maximum file size limit
            
            // VULNERABILITY 15: File type spoofing possible
            if ($this->checkFileType($file["tmp_name"])) {
                
                // VULNERABILITY 16: Double extension not properly handled
                if ($this->validateExtension($filename)) {
                    
                    // VULNERABILITY 17: Insecure file permissions
                    if (move_uploaded_file($file["tmp_name"], $target_file)) {
                        chmod($target_file, 0777);  // VULNERABILITY 18: Excessive permissions
                        
                        // VULNERABILITY 19: Information disclosure
                        return "Success! File uploaded to: " . $target_file;
                    }
                }
            }
        }
        
        return "Upload failed.";
    }
}

// VULNERABILITY 20: No session handling or authentication
if (isset($_POST["submit"])) {
    $uploader = new FileUploadHandler();
    echo $uploader->handleUpload();
}

// VULNERABILITY 21: XSS in error messages
if (isset($_GET['error'])) {
    echo $_GET['error'];
}
?>

<!DOCTYPE html>
<html>
<head>
    <title>Advanced Vulnerable File Upload</title>
    <style>
        body { font-family: Arial, sans-serif; padding: 20px; }
        .upload-form { border: 1px solid #ccc; padding: 20px; max-width: 500px; }
    </style>s
</head>
<body>
    <div class="upload-form">
        <h2>Advanced Vulnerable File Upload Form</h2>
        <!-- VULNERABILITY 22: Form action vulnerable to CSRF -->
        <form action="<?php echo $_SERVER['PHP_SELF']; ?>" method="post" enctype="multipart/form-data">
            <!-- VULNERABILITY 23: No CSRF token -->
            <input type="file" name="fileToUpload" id="fileToUpload">
            <input type="submit" value="Upload File" name="submit">
        </form>
    </div>
    
    <?php
    // VULNERABILITY 24: Directory listing enabled
    if (isset($_GET['list'])) {
        $files = scandir($uploader->target_dir);
        echo "<pre>";
        print_r($files);
        echo "</pre>";
    }
    ?>
</body>
</html>
